idpf_specification.md

```diff
@@ -6433,28 +6433,101 @@ VIRTCHNL2_OP_RDMA			= 529,
+8.	At this point the driver registers hooks into the stack for PTP runtime callbacks.
+9.	A driver may also request a separate dedicated mailbox for doing the indirect access such as for Tx Timestamping, adjusting the Device master clock etc. The way the Device control indicates to the driver to use a dedicated mailbox or not is again through the PTP capability negotiation data structures containing the flags for driver to do the right setup.
+
+#### Runtime flows
+Runtime flows are triggered by the stack and are dependent on what the system and the CPU Complex is being used for. If a system and the CPU Complex is used for Tenant hosting it may only have the Cross timestamping runtime flow. Whereas if a system and the CPU complex attached to the device is used for Infrastructure SW hosting then it may be used for adjusting the Device master clock directly or indirectly.
+If a system is used for a master Independent clock for the PTP network Domain, then it may have some subset of these runtime flows as desired from a given CPU Complex attached to the device. 
+##### Cross time stamping flow done on all CPU complexes attached to the device.
+1.	If the OS supports hooks for PCIE based PTM capability and a callback is received to do PTM cross time stamping, if the capability was granted to the driver instance, the driver makes a call to either do a direct or mailbox based PTM cross time stamping request to the device. Direct method is preferred over the mailbox method for cross timestamping to avoid unnecessary mailbox chatter. The opcodes used for indirect PTM access is VIRTCHNL2_OP_PTP_GET_CROSS_TIME
+2.	If the capability was not granted the driver will fail the call.
+3.	Legacy method of cross timestamping is pretty much supported in all modern OSes although may not be invoked in case of a VM. VM system time may be synchronized with the Hypervisor or Host OS time through a SW flow. Direct access is the only right method for this, mailbox mechanism is only for completeness. 
+##### Tx Timestamping Flow (given to drivers that are not untrusted(tenant) drivers.)
+a.	In this case if the device control plane grants the direct or indirect access for Tx timestamp, the driver learns about the number of latches and the index of the latches as part of initialization. If the access grated is direct access, the driver also learns about the register offsets of the latches in the PCIE MMIO space.
+b.	When a packet comes down with a PTP time stamping request as part of packet meta data from the stack then the driver must enable the Tx timestamping bits in the Tx descriptor and specify one of the free/unused latch ID to be used for timestamping of this Tx packet by the HW.
```

Comment:
'SKBTX_HW_TSTAMP' tx_flag in skb instead of 'PTP time stamping request'. There are multiple time stamp request types and not all of them are being handled.

Reply:
Again an OS /Linux detail that we don't mention in the Spec. As the spec is written in an OS agnostic way.